core/src/main/scala/com/metamx/tranquility/tranquilizer/Tranquilizer.scala

```diff
@@ -0,0 +1,310 @@
+    * MessageDroppedException, which means that a message was dropped due to unrecoverable reasons. With Druid this
+    * can be caused by message timestamps being outside the configured windowPeriod.
+    *
+    * NB: This can actually be wrong: if *any* message in a batch of messages is dropped, then all messages in that
+    * batch will be marked as dropped (even if they actually made it out). This is a limitation of how Tranquility
+    * tracks which messages were and were not dropped. This could be improved once
+    * https://github.com/druid-io/tranquility/issues/56 is addressed.
+    */
+  def send(message: MessageType): Future[Unit] = {
+    requireStarted()
+    if (log.isTraceEnabled) {
+      log.trace(s"Sending message: $message")
+    }
+    val holder = MessageHolder(message, Promise())
+
+    val (exBuffer, future) = lock.synchronized {
+      assert(buffer.size < maxBatchSize, "buffer.size < maxBatchSize")
+
+      buffer += holder
+
+      if (bufferStartMillis == 0) {
+        bufferStartMillis = System.currentTimeMillis()
+      }
+
+      val _exBuffer = if (buffer.size == maxBatchSize) {
+        Some(swap())
+      } else if (lingerMillis == 0 && pendingBatches < maxPendingBatches) {
+        Some(swap())
+      } else {
+        None
+      }
+
+      (_exBuffer, holder.future)
+    }
+
+    exBuffer foreach sendBuffer
+
+    future
+  }
+
+  /**
+    * Forces all pending messages to be sent immediately. Does not wait until they are sent; you still need to
+    * keep track of the returned futures.
+    */
+  def flush(): Unit = {
+    lock.synchronized {
+      if (buffer.nonEmpty) {
+        log.debug(s"Flushing ${buffer.size} buffered messages.")
+        sendBuffer(swap())
+      }
+
+      log.debug(s"Flushing $pendingBatches pending batches.")
+      while (pendingBatches > 0) {
+        lock.wait()
+      }
+    }
+  }
+
+  def simple(reportDropsAsExceptions: Boolean = false): SimpleTranquilizerAdapter[MessageType] = {
+    SimpleTranquilizerAdapter.wrap(this, reportDropsAsExceptions)
+  }
+
+  /**
+    * Stop the tranquilizer, and close the underlying Beam. Blocks until closed.
+    */
+  @LifecycleStop
+  def stop(): Unit = {
+    started = false
+    sendThread.interrupt()
+    sendThread.join()
+    Await.result(beam.close())
+  }
+
+  override def toString = s"Tranquilizer($beam)"
+
+  private def requireStarted() {
+    if (!started) {
+      throw new IllegalStateException("Not started")
+    }
+  }
+
+  // Swap out the current buffer, returning it and incrementing pendingBatches. Will block if more than
+  // maxPendingBatches are already pending.
+  private def swap(): Buffer[MessageHolder] = {
+    lock.synchronized {
+      while (pendingBatches >= maxPendingBatches) {
+        if (log.isDebugEnabled) {
+          log.debug(s"pendingBatches[$pendingBatches] >= maxPendingBatches[$maxPendingBatches], waiting...")
+        }
+        lock.wait()
```

Comment:
Its possible for multiple threads to wait here, after this all of them can start and swap the buffer, 
wondering if this can possibly lead to a case when a batch has 0 messages to send ? 
do we need any handling for sending empty buffer?